api/query/atoms.go

```diff
@@ -51,24 +51,53 @@ type AbstractExists struct {
 func (e AbstractExists) BindToRuns(runs ...shared.TestRun) ConcreteQuery {
 	queries := make([]ConcreteQuery, len(e.Args))
 	for i, arg := range e.Args {
-		byRun := make([]ConcreteQuery, 0, len(runs))
-		for _, run := range runs {
-			bound := arg.BindToRuns(run)
-			if _, ok := bound.(False); !ok {
-				byRun = append(byRun, bound)
+		var query ConcreteQuery
+		if _, isSeq := arg.(AbstractSequential); isSeq {
+			// For sequential, we pass all runs.
+			query = arg.BindToRuns(runs...)
+		} else {
+			// Everything else is split, one run must satisfy the whole tree.
+			byRun := make([]ConcreteQuery, 0, len(runs))
+			for _, run := range runs {
+				bound := arg.BindToRuns(run)
+				if _, ok := bound.(False); !ok {
+					byRun = append(byRun, bound)
+				}
 			}
+			query = Or{Args: byRun}
 		}
-		// Each separate Exists query is true if there exists a run that makes its concrete query true.
-		queries[i] = Or{
-			Args: byRun,
-		}
+		queries[i] = query
 	}
 	// And the overall node is true if all its exists queries are true.
 	return And{
 		Args: queries,
 	}
 }
 
+// AbstractSequential represents the root of a sequential queries, where the first
+// query must be satisfied by some run such that the next run, sequentially, also
+// satisfies the next query, and so on.
+type AbstractSequential struct {
+	Args []AbstractQuery
+}
+
+// BindToRuns binds each sequential query to an and-combo of those queries against
+// specific sequential runs, for each combination of sequential runs.
+func (e AbstractSequential) BindToRuns(runs ...shared.TestRun) ConcreteQuery {
+	numSeqQueries := len(e.Args)
+	byRuns := make([]ConcreteQuery, 0, len(runs)+1-numSeqQueries)
```

Comment:
Setting the capacity (the third argument) is only a minor performance optimization, not mandatory. Given how complicated this expression is, I'd say just omit it.